### PR TITLE
MINIFICPP-1535 - Support process groups

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by fgerlits** [comment]
Some unit tests with flows containing process groups would be useful.

File: libminifi/include/core/FlowConfiguration.h [hunk 1]
```diff
@@ -106,6 +106,7 @@ class FlowConfiguration : public CoreComponent {
   // Create Root Processor Group
 
   std::unique_ptr<core::ProcessGroup> createRootProcessGroup(std::string name, utils::Identifier & uuid, int version);
+  std::unique_ptr<core::ProcessGroup> createSimpleProcessGroup(std::string name, utils::Identifier & uuid, int version);
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/core/FlowConfiguration.h]: done

File: libminifi/src/core/ProcessGroup.cpp [hunk 2]
```diff
@@ -394,7 +391,7 @@ std::size_t ProcessGroup::getTotalFlowFileCount() const {
     sum += gsl::narrow<std::size_t>(conn->getQueueSize());
   }
 
-  for (const ProcessGroup* childGroup : child_process_groups_) {
+  for (auto& childGroup : child_process_groups_) {
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/core/ProcessGroup.cpp]: done

File: libminifi/src/core/yaml/YamlConfiguration.cpp [hunk 3]
```diff
@@ -49,35 +49,45 @@ YamlConfiguration::YamlConfiguration(const std::shared_ptr<core::Repository>& re
...
+std::unique_ptr<core::ProcessGroup> YamlConfiguration::createProcessGroup(YAML::Node yamlNode, bool is_root) {
+  utils::Identifier uuid;
+  int version = 0;
+
+  yaml::checkRequiredField(&yamlNode, "name", logger_,
+  CONFIG_YAML_REMOTE_PROCESS_GROUP_KEY);
+  std::string flowName = yamlNode["name"].as<std::string>();
+
+  std::string id = getOrGenerateId(&yamlNode);
   uuid = id;
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/core/yaml/YamlConfiguration.cpp]: done

File: libminifi/include/core/ProcessGroup.h [hunk 4]
```diff
@@ -57,7 +60,12 @@ enum ProcessGroupType {
 
 // ProcessGroup Class
 class ProcessGroup : public CoreComponent {
+  friend struct ::TestAccessor;
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/core/ProcessGroup.h]: renamed, the tests now use it

File: libminifi/src/core/FlowConfiguration.cpp [hunk 5]
```diff
@@ -116,6 +116,10 @@ std::unique_ptr<core::ProcessGroup> FlowConfiguration::createRootProcessGroup(st
   return std::unique_ptr<core::ProcessGroup>(new core::ProcessGroup(core::ROOT_PROCESS_GROUP, name, uuid, version));
 }
 
+std::unique_ptr<core::ProcessGroup> FlowConfiguration::createSimpleProcessGroup(std::string name, utils::Identifier & uuid, int version) {
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/core/FlowConfiguration.cpp]: done

File: libminifi/src/core/FlowConfiguration.cpp [hunk 6]
```diff
@@ -116,6 +116,10 @@ std::unique_ptr<core::ProcessGroup> FlowConfiguration::createRootProcessGroup(st
...
+  return std::unique_ptr<core::ProcessGroup>(new core::ProcessGroup(core::SIMPLE_PROCESS_GROUP, name, uuid, version));
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/core/FlowConfiguration.cpp]: done

File: libminifi/src/core/ProcessGroup.cpp [hunk 7]
```diff
@@ -87,17 +87,17 @@ ProcessGroup::~ProcessGroup() {
...
 }
 
 bool ProcessGroup::isRootProcessGroup() {
-  std::lock_guard<std::recursive_mutex> lock(mutex_);
   return (type_ == ROOT_PROCESS_GROUP);
 }
 
+bool ProcessGroup::isRemoteProcessGroup() {
+  return (type_ == REMOTE_PROCESS_GROUP);
+}
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/core/ProcessGroup.cpp:98]: done

File: libminifi/src/core/ProcessGroup.cpp [hunk 8]
```diff
@@ -255,21 +245,21 @@ void ProcessGroup::stopProcessing(const std::shared_ptr<TimerDrivenSchedulingAge
   }
 }
 
-std::shared_ptr<Processor> ProcessGroup::findProcessorById(const utils::Identifier& uuid) const {
+std::shared_ptr<Processor> ProcessGroup::findProcessorById(const utils::Identifier& uuid, Traverse traverse) const {
   const auto id_matches = [&] (const std::shared_ptr<Processor>& processor) {
     logger_->log_debug("Current processor is %s", processor->getName());
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/core/ProcessGroup.cpp]: done

File: libminifi/src/core/yaml/YamlConfiguration.cpp [hunk 9]
```diff
@@ -49,35 +49,45 @@ YamlConfiguration::YamlConfiguration(const std::shared_ptr<core::Repository>& re
...
     }
   }
 
-  std::string id = getOrGenerateId(&rootFlowNode);
+  auto rootGroup = parseProcessGroupYaml(flowControllerNode, rootFlowNode, true);
+  this->name_ = rootGroup->getName();
+  return rootGroup;
+}
+
+std::unique_ptr<core::ProcessGroup> YamlConfiguration::createProcessGroup(YAML::Node yamlNode, bool is_root) {
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/core/yaml/YamlConfiguration.cpp]: changed all to use `const&`

**Comment by adamdebreceni** [comment]
> Some unit tests with flows containing process groups would be useful.

I have restored the mixed up ProcessGroup parsing unit tests (accidentally committed them on a different branch 😄), should we create some tests that actually trigger a flow?

File: extensions/standard-processors/tests/unit/ProcessGroupTestUtils.h [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+#pragma once
+
+#include "TestBase.h"
+#include "YamlConfiguration.h"
+#include "Utils.h"
+
+struct Lines {
+  std::vector<std::string> lines;
+
+  std::string join(const std::string& delim) const {
```
Comment (fgerlits) [inline extensions/standard-processors/tests/unit/ProcessGroupTestUtils.h:27]: this seems to be the same as `StringUtils::join(const std::basic_string<TChar>& separator, const U& container)`

File: extensions/standard-processors/tests/unit/YamlProcessGroupParserTests.cpp [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+      .With({RPG{"Child1_RPG1", {
+        Proc{"Port1", "00000000-0000-0000-0000-000000000005"}}}})
+    });
+
+  auto root = config.getRootFromPayload(pattern.serialize().join("\n"));
+
+  verifyProcessGroup(*root, pattern);
+}
+
+TEST_CASE("Cannot connect processors from different groups", "[YamlProcessGroupParser3]") {
```
Comment (fgerlits) [inline extensions/standard-processors/tests/unit/YamlProcessGroupParserTests.cpp:73]: The name of this test is misleading, as it verifies that you _can_ connect processors from different groups.

**Review by fgerlits — COMMENTED** [review]
> I have restored the mixed up ProcessGroup parsing unit tests (accidentally committed them on a different branch smile), should we create some tests that actually trigger a flow?

These yaml parsing tests are enough, I think.  They are very nice.

File: extensions/standard-processors/tests/unit/ProcessGroupTestUtils.h (hunk 10 above)
Comment (adamdebreceni) [reply extensions/standard-processors/tests/unit/ProcessGroupTestUtils.h:27]: done

File: extensions/standard-processors/tests/unit/YamlProcessGroupParserTests.cpp (hunk 11 above)
Comment (adamdebreceni) [reply extensions/standard-processors/tests/unit/YamlProcessGroupParserTests.cpp:73]: this test was quite cryptic, I have changed `MaybeProc{"<id>", utils::nullopt}` to `UnresolvedProc{"<id>"}`, does this make it easier to understand?

Comment (fgerlits) [reply extensions/standard-processors/tests/unit/YamlProcessGroupParserTests.cpp:73]: Yes, this helps.  I forgot that a null optional for the ID means "verify that this processor is not found".  It is still a bit tricky, but clearer.

**Review by fgerlits — APPROVED** [review]

**Review by lordgamez — APPROVED** [review]

**Pull request closed without merging**